sqlglot/dataframe/sql/dataframe.py

```diff
@@ -0,0 +1,737 @@
+                    if matching_ctes:
+                        for matching_cte in matching_ctes:
+                            if matching_cte.alias_or_name in [join_table.alias_or_name for join_table in join_tables]:
+                                sequence_id_expression.set("this", matching_cte.args["alias"].this)
+                                df.pending_hints.remove(hint)
+                                break
+                hint_expression.args.get("expressions").append(hint)
+        if hint_expression.expressions:
+            expression.set("hint", hint_expression)
+        return df
+
+    def _hint(self, hint_name: str, args: t.List[Column]) -> DataFrame:
+        hint_name = hint_name.upper()
+        if hint_name in JOIN_HINTS:
+            hint_expression = exp.JoinHint(
+                this=hint_name, expressions=[exp.to_table(parameter.alias_or_name) for parameter in args]
+            )
+        else:
+            hint_expression = exp.Anonymous(this=hint_name, expressions=[parameter.expression for parameter in args])
+        new_df = self.copy()
+        new_df.pending_hints.append(hint_expression)
+        return new_df
+
+    def _set_operation(self, clazz: t.Callable, other: DataFrame, distinct: bool):
+        other_df = other._convert_leaf_to_cte()
+        base_expression = self.expression.copy()
+        base_expression = self._add_ctes_to_expression(base_expression, other_df.expression.ctes)
+        all_ctes = base_expression.ctes
+        other_df.expression.set("with", None)
+        base_expression.set("with", None)
+        operation = clazz(this=base_expression, distinct=distinct, expression=other_df.expression)
+        operation.set("with", exp.With(expressions=all_ctes))
+        return self.copy(expression=operation)._convert_leaf_to_cte()
+
+    def _cache(self, storage_level: str):
+        df = self._convert_leaf_to_cte()
+        df.expression.ctes[-1].set("cache_storage_level", storage_level)
+        return df
+
+    @classmethod
+    def _add_ctes_to_expression(cls, expression: exp.Expression, ctes: t.List[exp.CTE]) -> exp.Expression:
+        expression = expression.copy()
+        with_expression = expression.args.get("with")
+        if with_expression:
+            existing_ctes = with_expression.args["expressions"]
+            existsing_cte_names = [x.alias_or_name for x in existing_ctes]
+            for cte in ctes:
+                if cte.alias_or_name not in existsing_cte_names:
+                    existing_ctes.append(cte)
+        else:
+            existing_ctes = ctes
+        expression.set("with", exp.With(expressions=existing_ctes))
+        return expression
+
+    @classmethod
+    def _get_outer_select_columns(cls, item: t.Union[exp.Expression, DataFrame]) -> t.List[Column]:
+        expression = item.expression if isinstance(item, DataFrame) else item
+        return [Column(x) for x in dict.fromkeys(expression.find(exp.Select).args.get("expressions", []))]
```

Comment:
are you trying to deduplicate this?

Reply:
find(exp.Select).expressions, automatically gives you the empty list default

Reply:
Yeah I think this can actually be removed. Thanks for calling it out.